### PR TITLE
🥅 coerce and trap UTF-8 user input

### LINKED ISSUE #143: 🚸 strip quotes from search criteron
A search like "blah" should be sought as blah.  The same should apply to 'blah',
and any UTF-8 should also be stripped.

### PR DESCRIPTION
**Pull request by daybologic**
search detainting / stripping " and ', coercion from modern quotes back to ASCII.

Basic UTF-8 decontamination.

Ready for merge.

### Summary by CodeRabbit

The changes introduced in the code seem to be primarily focused on improving security, particularly input sanitization. The introduction of `Chleb::Utils::SecureString` module and its usage across different modules is a good step towards ensuring that user inputs are properly sanitized before being processed.

Here are some suggestions for further improvements:

1. **Error Handling**: It's not clear from the summary how errors are handled during the detainting process. If an error occurs (for example, if the input string contains illegal characters), it would be beneficial to have a robust error handling mechanism in place. This could include throwing exceptions or returning error codes/messages that can be logged and/or displayed to the user.

2. **Performance**: Depending on the size of the input strings and the frequency of calls to the detainting functions, there might be performance implications. Consider benchmarking these operations to ensure they don't significantly impact the application's performance.

3. **Testing**: Ensure that there are sufficient tests covering all possible edge cases for the new `Chleb::Utils::SecureString` module. This includes testing with various types of input - valid strings, strings with illegal characters, empty strings, etc.

4. **Consistency**: In `Chleb::Server::Dancer2.pm`, you're replacing direct calls to `param()` with calls to `__param()`. Make sure this change is consistently applied throughout the codebase to avoid confusion and potential bugs.

5. **Maintainability**: The `__ensureSecureString` function in `Chleb.pm` seems to be doing similar work as the `__detaint` function in other modules. Consider consolidating these into a single utility function/module to reduce redundancy and improve maintainability.

6. **Security**: In `lib/Chleb/Server/Moose.pm`, URI escaping is used for query parameters. While this is generally a good practice, it's important to ensure that all parts of the URI are properly escaped, not just the query parameters. Also, consider using a more modern and secure library for URI encoding/decoding if available.

7. **Best Practices**: The use of `MooseTrimmedStr` subtype in `Chleb::Bible::Search::Query.pm` is a good practice as it ensures type safety. However, make sure that this doesn't introduce unnecessary complexity or performance overhead.

### CODE REVIEW

File: lib/Chleb/Utils/SecureString.pm [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+			}
+		}
+	}
+
+	return __PACKAGE__->new({
+		stripped => $stripped,
+		tainted  => 0,
+		value    => $detaintedValue,
+	});
+}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Utils/SecureString.pm:349]: have since changed the code, so might need a new suggestion

File: lib/Chleb/Utils/SecureString.pm [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+	return unless ($mode);
+	return if ($mode == $MODE_TRAP);
+	return if ($mode == $MODE_PERMIT);
+
+	return die Chleb::Exception->raise(
+		HTTP_INTERNAL_SERVER_ERROR,
+		'Illegal mode in call to Chleb::Utils::SecureString/detaint',
+	);
+}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Utils/SecureString.pm]: 3e9bdc5088badf795b80db44c21aa4986afb53dc

File: lib/Chleb.pm [hunk 3]
```diff
@@ -78,11 +78,14 @@ sub newSearchQuery {
...
 		($defaults{bible}) = $self->__getBible();
-		return Chleb::Bible::Search::Query->new({ %defaults, text => $args[0] })
+		return Chleb::Bible::Search::Query->new({ %defaults, text => Chleb::Utils::SecureString->new({ value => $args[0] }) })
 	}
 
 	my %params = @args;
 	__fixTranslationsParam(\%params);
+	if (ref($params{text}) eq 'Chleb::Utils::SecureString') {
+		$params{text} = Chleb::Utils::SecureString->new({ value => $params{text} })
+	}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb.pm]: 670a168

File: lib/Chleb.pm [hunk 4]
```diff
@@ -78,11 +78,14 @@ sub newSearchQuery {
 
 	if (scalar(@args) == 1) {
 		($defaults{bible}) = $self->__getBible();
-		return Chleb::Bible::Search::Query->new({ %defaults, text => $args[0] })
+		return Chleb::Bible::Search::Query->new({ %defaults, text => Chleb::Utils::SecureString->new({ value => $args[0] }) })
 	}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb.pm:82]: That's not how this works, detaint is called later, when the search is executed.  It cannot return undef here.

File: lib/Chleb.pm [hunk 5]
```diff
@@ -78,11 +78,14 @@ sub newSearchQuery {
...
 
 	my %params = @args;
 	__fixTranslationsParam(\%params);
+	if (ref($params{text}) ne 'Chleb::Utils::SecureString') {
+		$params{text} = Chleb::Utils::SecureString->new({ value => $params{text} })
+	}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb.pm]: incorrect interpretation of the logic

File: lib/Chleb/Utils/SecureString.pm [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+	}
+
+	return 0;
+}
+
+=back
+
+=cut
+
+1;
```
Comment (daybologic) [inline lib/Chleb/Utils/SecureString.pm:376]: this doesn't work, not sure why.  Will push in a new PR when I figure out why

File: t/peace_on_earth.t [hunk 7]
```diff
@@ -148,7 +166,7 @@ sub testPeaceSearch_defaultTranslation {
 		),
 	), 'results inspection');
 
-	return EXIT_SUCCESS;
+	return;
```
[the remark this answers is not in the document]
Comment (daybologic) [reply t/peace_on_earth.t:169]: It is unnecessary because it's not a method called by the test framework directly any more.

File: lib/Chleb/Server/Moose.pm [hunk 8]
```diff
@@ -608,7 +608,7 @@ sub __search {
 		},
 	);
 
-	$hash{links}->{self} = '/1/search?term=' . $search->{term} . '&wholeword=' . $wholeword .'&limit=' . $limit;
+	$hash{links}->{self} = '/1/search?term=' . $query->text . '&wholeword=' . $wholeword .'&limit=' . $limit;
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Server/Moose.pm]: 6303b90d8f436812573e0e10c8cca0a5bf15e5a7

File: lib/Chleb/Server/Dancer2.pm [hunk 9]
```diff
@@ -149,6 +150,29 @@ sub __configGetPublicDir {
...
+
+	return $detainted;
+}
+
+sub __param {
+	my ($name) = @_;
+	my $value = param($name);
+	return undef unless (defined($value));
+	return __detaint($value, $name);
+}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Server/Dancer2.pm:174]: No, __detaint handles that within SecureString, with a nice error, and it throws the exception.  We can't put undef checks in everywhere, it makes the code messy.

File: t/peace_on_earth.t [hunk 10]
```diff
@@ -59,9 +60,26 @@ sub setUp {
...
+	return EXIT_SUCCESS;
+}
+
+sub testPeaceSearch_defaultTranslation_viaObject {
+	my ($self) = @_;
+
+	my $secureString = Chleb::Utils::SecureString->new({ value => 'peace on earth' });
+	$self->__checkPeaceSearch_defaultTranslation(text => $secureString);
+
+	return EXIT_SUCCESS;
```
[the remark this answers is not in the document]
Comment (daybologic) [reply t/peace_on_earth.t:75]: that would break the meaning of the test wrappers, the point is to make sure it works both ways, if we always constructed the object, the logic wouldn't be tested in the library.

File: lib/Chleb.pm (hunk 5 above)
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb.pm]: 0822d8c

File: lib/Chleb/Bible/Search/Query.pm [hunk 11]
```diff
@@ -32,17 +32,24 @@ package Chleb::Bible::Search::Query;
...
+use Chleb::Utils::SecureString;
 use Readonly;
 use Time::HiRes ();
 
 Readonly our $SEARCH_RESULTS_LIMIT => 50;
 
+subtype 'MooseTrimmedStr', as 'Str';
+coerce 'MooseTrimmedStr', from 'Chleb::Utils::SecureString', via {
+	__detaint($_)
+};
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Bible/Search/Query.pm:51]: this is interesting... but it works how it is.  I might address this in bugfix/strip-quotes-and-utf8/2

File: lib/Chleb/Bible/Search/Query.pm [hunk 12]
```diff
@@ -114,6 +121,16 @@ sub toString {
...
 
+sub __detaint {
+	my ($value) = @_;
+	my $mode = $Chleb::Utils::SecureString::MODE_TRAP
+	    | $Chleb::Utils::SecureString::MODE_COERCE
+	    | $Chleb::Utils::SecureString::MODE_STRIP_QUOTES
+	    | $Chleb::Utils::SecureString::MODE_TRIM;
+
+	return Chleb::Utils::SecureString::detaint($value, $mode, 'term')->value;
+}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Bible/Search/Query.pm:132]: Who says it's better?  Putting all the method declarations above the private methods make them more accessible for readers of man pages and generally speaking anybody looking at the code.  This advice sounds like it came from single-pass compilers or something.  We have moved on.

File: lib/Chleb/Utils/SecureString.pm (hunk 6 above)
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Utils/SecureString.pm:376]: I don't think point 2 is true.  it's either coerced, or we check if it's in range.  If this is inefficient, we might need to address it in bugfix/strip-quotes-and-utf8/2 now, as this feedback loop is getting a bit long.

File: lib/Chleb.pm [hunk 13]
```diff
@@ -78,11 +78,14 @@ sub newSearchQuery {
...
 		($defaults{bible}) = $self->__getBible();
-		return Chleb::Bible::Search::Query->new({ %defaults, text => $args[0] })
+		return Chleb::Bible::Search::Query->new({ %defaults, text => Chleb::Utils::SecureString->new({ value => $args[0] }) })
 	}
 
 	my %params = @args;
 	__fixTranslationsParam(\%params);
+
+	$params{text} = __ensureSecureString($params{text});
+
```
Comment (daybologic) [inline lib/Chleb.pm:88]: 05556b5

[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb.pm:88]: No, I don't want to take this apart in this way.  When the query executes, it will detaint the SecureString (text) and it's at that point we will throw the appropriate exception.

File: lib/Chleb.pm [hunk 14]
```diff
@@ -400,6 +403,16 @@ sub __isTestamentMatch {
...
 
+sub __ensureSecureString {
+	my ($input) = @_;
+
+	if (!defined($input) || ref($input) ne 'Chleb::Utils::SecureString') {
+		return Chleb::Utils::SecureString->new({ value => $input });
+	}
+
+	return $input;
+}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb.pm:414]: It's when the SecureString is detained, that any exception will be thrown
